snapcraft/tests/unit/project_loader/grammar_processing/test_part_grammar_processor.py

```diff
@@ -14,13 +14,58 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+import snapcraft
+from snapcraft.tests import unit
+from testtools.matchers import Equals
+
 from snapcraft.internal.project_loader.grammar_processing import (
+    PartGrammarProcessor,
     _part_grammar_processor as processor
 )
 
 import doctest
+from unittest import mock
 
 
 def load_tests(loader, tests, ignore):
     tests.addTests(doctest.DocTestSuite(processor))
     return tests
+
+
+class PartGrammarTestCase(unit.TestCase):
+
+    scenarios = [
+        ('plain string', {
+            'properties': {'source': 'foo'},
+            'target_arch': 'amd64',
+            'expected': {'source': 'foo'}
+        }),
+        ('on amd64', {
+            'properties': {'source': [{'on amd64': 'foo'}]},
+            'target_arch': 'amd64',
+            'expected': {'source': 'foo'}
+        }),
+        ('on i386', {
+            'properties': {'source': [{'on i386': 'foo'}, {'else': 'bar'}]},
+            'target_arch': 'amd64',
+            'expected': {'source': 'bar'}
+        }),
+        ('on i386, target_arch=i386', {
+            'properties': {'source': [{'on i386': 'foo'}, {'else': 'bar'}]},
```

Comment:
can you add more values than just the ones that need modification to verify the remain intact?